Auction House Overview (#1)
0xproflupin/MetaplexDocs: AuctionHouse/Overview.md

# Auction House: Overview

## Introduction
Auction House is a program that allows users to exchange assets within the Solana blockchain. 

> There are plenty of ways to exchange assets on Solana, so why another protocol focused on solving this problem? Let's dive deep into that.

Reviewer: We can call programs any things (protocol, smart contract, etc.) but let's keep a consistent language here and stick with "program" everywhere.

```suggestion
There are plenty of ways to exchange assets on Solana, so why another program focused on solving this problem? Let's dive deep into that.
```

The most important aspect of the Auction House protocol is that it provides ownership of assets to the user.

Traditionally, as soon as the user lists an asset on a marketplace, the asset is moved from the user's wallet into a wallet known as the [Escrow](https://www.investopedia.com/terms/e/escrow.asp) wallet owned by the marketplace and is kept or **escrowed** there until the asset is delisted or sold. This poses some concerns:
* The user can not list the same asset on multiple marketplaces
* The user has to rely on the marketplaces’ escrow contracts to safely hold their asset.

This is where Auction House shows its power. Its a transaction protocol that allows marketplaces to implement an **escrow-less** sales contract, thus providing ownership of assets to the users.

## Creating an Auction House

> The Auction House program can be used to make a marketplace by creating a new **Auction House Instance**. The `Auction House Instance` is a [Program Derived Address (PDA)](https://solanacookbook.com/core-concepts/pdas.html) which is in the keyspace of the Auction House program, ie, the `Instances` are simply public keys owned by the Auction House program.

Reviewer: - Let's keep the "`" symbol for code-related text.
- Let's keep a consistent language regardings accounts, instructions and programs.
- Link to our internal documentation when possible since we already list additional resources in there.
- The "keyspace of the Auction House program" part is confusing.

```suggestion
The Auction House program can be used to create a new marketplace by instantiating a new **Auction House** account. The Auction House account is a [Program Derived Address (PDA)](/programs/understanding-programs#program-derived-addresses-pda) which is derived from a given public key and, optionally, an SPL Token to use a currency (more on that below).
```

![](https://i.imgur.com/2HPpM9g.png =300x)

> The Instance can be configured in the whatever way the user wants. Here are some important configurable parameters:

Reviewer: ```suggestion
The account can be configured in whichever way the user wants. We'll talk [more about these configurations in a dedicated page](/TODO/when-we-the-page-is-written) but here are some interesting configurable parameters:
```

> * `requireSignOff`: this allows marketplaces to gate which asset can be listed, bought and sold. For every action: list, bid, sale, the instance [authority](https://docs.solana.com/staking/stake-accounts#understanding-account-authorities) needs to sign the transaction

Reviewer: - "bought and sold" is confusing because they are synonyms outside of this context.
 
```suggestion
* `requireSignOff`: this allows marketplaces to gate which assets can be listed and which bids can be placed. On every relevant instruction, the Auction House [authority](https://docs.solana.com/staking/stake-accounts#understanding-account-authorities) needs to sign the transaction.
```

> * `canChangeSalePrice`: `canChangeSalePrice` is only intended to be used with AuctionHouses with `requireSignOff` set to `true`. This allows the Auction House to do complicated order matching to find the best price for the seller.

Reviewer: ```suggestion
* `canChangeSalePrice`: this parameter is only intended to be used on Auction Houses with `requireSignOff` set to `true`. This allows the Auction House to perform custom order matching to find the best price for the seller.
```

> * `sellerFeeBasisPoints`: the share of the sale the marketplace takes on all trades

Reviewer: ```suggestion
* `sellerFeeBasisPoints`: this represents the share the marketplace takes on all trades. For instance, if this is set to `200`, i.e. 2%, then the marketplace takes 2% of every single trade that happens on their platform.
```

    
## Listing and Bidding
Once the marketplace is active, any trader can post an offer to "sell" an asset on the marketplace. When this happens, the Auction House does two things:

1. Auction House creates a `Sell Order`: in other words, creates a PDA known as the `SellerTradeState` which represents the offer. Trade States are special PDAs that are very cheap in comparison to other PDAs / Accounts. This is because these PDAs only have 1 byte of data, which is the bump of the PDA. All the details of a listings such as list price, amount of tokens, mint account etc, are hashed into the seeds of the PDA and therefore the PDA acts as a "proof of existence" for that listing.

![](https://i.imgur.com/ki27Ds8.png)

2. Auction House also assigns the another PDA: `programAsSigner` PDA as the `Delegate`. Delegates are a feature of the Solana SPL-token program and are discussed in detail [here](https://spl.solana.com/token#authority-delegation). Delegation allows the Auction House Instance to pull assets out of a token account when a sale goes through at a later point.

![](https://i.imgur.com/aIRl7Hb.png)

> Similar to the `Sell Order`, when a buyer places a bid, Auction House creates the `BuyerTradeState` PDA representing the bid. The bid amount (native or SPL tokens) then needs to be transferred manually to the `BuyerEscrowAccount` PDA, which holds this amount till the sale goes through.

Reviewer: There's too much different lingo here. I think we should focus on using:
- To list / A listing: for listing an asset.
- To bid / A bid: for placing a bid on an asset.

In this section, we just need to focus on these two primitive: listing an asset and placing a bid on an asset. You're doing this but the language makes it confusing because we're using words like selling and buying at this point, when actually, buying and selling is a combination of these two primitives with "execute sale" which is explained in the next section.

I'd love to see a more explicit explanation of these two primitives with an example that helps the reader visualise what we're doing.

    
## Executing a Sale
As soon as the buyer and the seller have both posted there bids and offers respectively, the trade can be completed by calling the `executeSale` instruction, which is a permission-less crank: in other words, can be executed by anyone without any fee or reward.

The `BuyerEscrowAccount` will transfer the tokens to the seller (minus Auction House fees). The seller's token account will send the asset to the buyer via a delegate transfer, completing a trade.
![](https://i.imgur.com/gpAX63m.png)

There are three trade scenarios in which the `executeSale` instruction may be called in different ways:

* **Bid + Sale in same transaction**: Alice lists an asset, Bob matches the offer and the sale goes through immediately.
* **Offer + Sale in same transaction**: Bob makes an offer of 5 SOL for an unlisted asset owned by Alice, Alice likes the offer and lists the asset for 5 SOL thus completing the sale immediately.

> * **Offer, Bid, Sale in three separate instructions**: Alice lists an asset for 5 SOL, Bob makes an offer for this asset for 3 SOL, Alice likes the offer and agrees to the sale.

Reviewer: This is a good attempt at using the primitives explained in the previous section to explain additional concepts but the reader gets a bit lost.

This could do with:
- A clearer distinction between explanation and examples.
- A more explicit distinction between the parts you're trying to explain.
  - First, the "execute sale" primitive.
  - Then, the concept of "buying" by placing a bid and executing the sale in the same transaction for an existing listing.
  - Finally, the concept of "selling" by listing and executing the sale in the same transaction for an existing bid.
- A simple diagram for each scenario.

    
## Auctioning Fungible Assets

> An asset can be any SPL token and doesn't have to be an NFT ("Non-Fungible Token"). 
> 
> The only constraint is that the Mint Account of the SPL Token need to have a Metadata account attached to it as a PDA. You can read more about this in the [Token Metadata program documentation](https://docs.metaplex.com/programs/token-metadata/overview).

Reviewer: From this point forward, the transitions between sections are more direct. It would be nice to add a little sentence that helps the reader continue in the storyline. For example, here we could say.

 ```suggestion
So far, we've talked about exchanging assets using an Auction House account, but we've not dug into what type of assets can be exchanged that way. The most popular assets that can be listed in an Auction House are [Non-Fungible Tokens (NFTs)](/resources/definitions#non-fungible-tokens).

However, these are not the only assets that can benefit from the Auction House program. In fact, an asset can be any SPL Token so long as it has a Metadata account attached to its Mint account. If you'd like to know more about SPL Token and Metadata accounts, you can [read more about this in the overview of our Token Metadata program](/programs/token-metadata/overview).
```

Some of the other sections could do with a similar update to guide the reader a bit better.

## Auctioneer
The current Auction House implementation is designed with instant sales in mind and currently has no features that enable the various auction types that have become popular in the Solana ecosystem (ex. ***Timed auctions***) 

Auctioneer is a customized contract type, written by the user, that uses the composability pattern of Auction House to control an individual Auction House instance. 

To fully enable Auctioneer to use an Auction House instance's instructions, it must be explicitly delegated. 

![](https://i.imgur.com/RyZUfR9.png)

## Next steps
In this page we have gone through the very basics of the Auction House protocol and the power it possesses. There is whole lot more Auction House is capable of and a lot more that can be configured to your liking.

To get a good understanding of how the Auction House works, we first need to dive a little deep into the Accounts and the Instructions that are are a part of the Auction House. Kindly go through these pages to get a good unserstanding of the above:
* [Getting Started]()
* [Accounts]()
* [Instructions]()
* [Executing Sales]()

> * [Cancelling Bids and Listings]()

Reviewer: I think we should introduce "Getting Started", "Accounts" and "Instructions" separately to the other two. The formers are repeated on every program documentation and act as human-friendly API references whereas the latters are feature pages specific to Auction House.

# Auction House: Overview

## Introduction
Auction House is a program that allows users to exchange assets within the Solana blockchain. 

There are plenty of ways to exchange assets on Solana, so why another program focused on solving this problem? Let's dive deep into that.

The ethos of this program is to allow anyone to create and configure their own marketplace and even provide their own custom logic on how assets should be exchanged. The motivation behind the Auction House protocol is to create a healthy ecosystem of marketplaces that focus on different use-cases, and more importantly, each bringing their own flavor into the way they allow users to trade assets.

> The most important aspect of the Auction House protocol is that it provides ownership of assets to the user.

Reviewer: ```suggestion
The most important aspect of the Auction House program is that it provides ownership of assets to the user.
```

Maybe do a cmd+f to be sure we didn't miss any.

* The user can not list the same asset on multiple marketplaces
* The user has to rely on the marketplaces’ escrow contracts to safely hold their asset.

This is where Auction House shows its power. Its a transaction protocol that allows marketplaces to implement an **escrow-less** sales contract, thus providing ownership of assets to the users.

## Creating an Auction House
The Auction House program can be used to create a new marketplace by instantiating a new **Auction House** account. The Auction House account is a [Program Derived Address (PDA)](/programs/understanding-programs#program-derived-addresses-pda) which is derived from a given public key and, optionally, an SPL Token to use a currency (more on that below).

![](https://i.imgur.com/2HPpM9g.png =300x)

The account can be configured in whichever way the user wants. We'll talk [more about these configurations in a dedicated page](/TODO/when-we-the-page-is-written) but here are some interesting configurable parameters:

> * `requireSignOff`: this allows marketplaces to gate which assets can be listed and which bids can be placed. On every relevant instruction, the Auction House [authority](https://docs.solana.com/staking/stake-accounts#understanding-account-authorities) needs to sign the transaction.8

Reviewer: ```suggestion
* `requireSignOff`: this allows marketplaces to gate which assets can be listed and which bids can be placed. On every relevant instruction, the Auction House [authority](https://docs.solana.com/staking/stake-accounts#understanding-account-authorities) needs to sign the transaction.
```

![](https://i.imgur.com/aIRl7Hb.png)

### Bidding
Similar to the case of listing, when a user places a bid, the Auction House creates a **Buy Order**, in other words, creates the `BuyerTradeState` PDA representing the bid. The bid amount (native or SPL tokens) needs to be transferred manually by the marketplace to the `BuyerEscrowAccount` PDA, which holds this amount till the sale goes through.

> Example:
> * Alice lists an asset A for 5 SOL. In doing so, the Auction House creates the `SellerTradeState` PDA representing the bid. The Auction House also assigns the `programAsSigner` PDA as the **Delegate**, hence giving it the **Authority** to pull the asset from Alice's wallet when the sale goes through.
> * Bob places a bid of 3 SOL on asset A. In doing so, the marketplace pulls 3 SOL from Bob's wallet to the `BuyerEscrowAccount` PDA. This amount will stay here up until the sale goes through.
    
## Executing a Sale

> Once we have a listing and a at least one bid placed for a given asset, a trade can be completed by calling the `executeSale` instruction. 

Reviewer: ```suggestion
Once we have a listing and at least one bid placed for a given asset, a trade can be completed by calling the `executeSale` instruction. 
```

* The Auction House pulls the bid amount stored in the `BuyerEscrowAccount` and transfers this amount to the lister (minus Auction House fees). 
* The `programAsSigner` PDA, which the Auction House assigned as the **Delegate**, pulls the asset from the lister's wallet (more specifically, from the Token Account in the lister's wallet), and transfers the asset into the bidder's wallet, thus completing the trade.
![](https://i.imgur.com/gpAX63m.png)

Now that we know how `executeSale` instruction works, lets discuss the three trade scenarios in which the `executeSale` instruction is executed in different ways:

1. *Lister agreeing to a bid*: This is the case when the `executeSale` instruction is executed independently, after a **Buy Order** and a **Sell Order** exist for a given asset.

> Example:
> * Alice lists an asset A for 5 SOL, thus the Auction House creates a **Sell Order** for the asset A.
> * Bob places a bid of 3 SOL for the asset A. The Auction House creates a **Buy Order** for the asset A.

> > * Alice agrees to the offer, thus enabling the marketplace to "execute the sale" using the `executeSale` instruction.

Reviewer: This example is weird because the listing and the bid don't match so actually, Alice would have to create a sell order a 3 SOL first before executing the sale which is equivalent to scenario 3 in this section.

I'd replace this with something like:
- Alice lists an asset for 5 SOL.
- Bob place a bid for 5 SOL on that asset unaware of Alice's listing.
- Alice sees that there's a matching bid for her listing and executes the sale.

Since this is more of a nice example, I'd put it as the last scenario just to explain that the execute sale can be independent of the other two primitives.

> 2. *"Buying" at list price*: This is the case when the bidder places a bid for an already listed asset, at the list amount itself. In such cases, the `executeSale` instruction is executed in the same transaction in which the bid is made (the **Buy Order** is made), and thus the bidder "buys" the asset.

Reviewer: Could you simplify this paragraph? It's a bit too technical and I've had to read it several times to grasp the information.

> Example:
> * Alice lists an asset A for 5 SOL, thus the Auction House creates a **Sell Order** for the asset A.
> * Bob places a bid of 5 SOL for the asset A, creating a **Buy Order** for the asset A.
> * This enables the marketplace to place a bid on the asset and execute the sale in the same transaction, in practice allowing Bob to "buy" asset A.

> 3. *"Selling" at bid price*: This is the case when a user interested in an unlisted asset places a bid on it. Once this happens, if the asset owner, now interested in selling the asset, lists the asset for the bid amount, the `executeSale` instruction is executed in the same transaction in which the listing is done (the **Sell Order** is made), and thus the lister "sells" the asset.

Reviewer: Could you simplify this paragraph? It's a bit too technical and I've had to read it several times to grasp the information.

> Example:
> * Bob places a bid of 5 SOL for an unlisted asset A. 
> * Alice notices the bid and lists the asset A for 5 SOL, creating a **Sell Order** for the asset A. 
> * This enables the marketplace to list the asset and execute the sale in the same transaction, in practice allowing Alice to "sell" asset A.
    
## Auctioning Fungible Assets
So far, we've talked about exchanging assets using an Auction House account, but we've not dug into what type of assets can be exchanged that way. The most popular assets that can be listed in an Auction House are [Non-Fungible Tokens (NFTs)](/resources/definitions#non-fungible-tokens).

However, these are not the only assets that can benefit from the Auction House program. In fact, an asset can be any SPL Token so long as it has a Metadata account attached to its Mint account. If you'd like to know more about SPL Token and Metadata accounts, you can [read more about this in the overview of our Token Metadata program](/programs/token-metadata/overview).
    
## Buying asset using a custom SPL Token

> In all the examples showcased above, we use SOL as the exchange currency to discuss how the Auction House program works. But SOL is not the only currency that can be configured to be used for exchanging assets and the Auction House program allows the marketplaces to configure any SPL-token to act as the currency for exchanging assets on their platform.

Reviewer: This is very good!

The sentences are a bit too long though. How about...

```suggestion
In the examples showcased above, we've used SOL as the exchange currency to discuss how the Auction House program works. But SOL is not the only currency that can be configured for exchanging assets. The Auction House program allows marketplaces to configure any SPL-token to act as their currency.
```

I think in general, your sentences are very good but good have much more impact by simplifying them slightly. Remove a few redundant words so readers get to the point quicker. Insert smaller sentences here and there to control the flow. Allow them to breathe. See what I did there. 😅

Author: Completely understand your point, that's a great suggestion Loris. I can notice I've used words that are causing pleonasm in the sentences. Will make sure to cut these off :)

This can be achieved by setting the `treasuryMint` parameter in the Auction House account to the mint account of the SPL-token of your liking.

## Custom Order Matching
When we were discussing Trade States, there was a third Trade State which was shown in the Trade State diagram: the `FreeSellerTradeState`. What's the utility of this Trade State?

> During the introduction to the Auction House program, we briefly discussed how Auction House can be used by marketplaces to provide their own custom logic on how assets should be exchanged. This is where the `FreeSellerTradeState` comes in.

Reviewer: Very nice!

When a buyer intentionally lists their asset for a price of 0 SOL / SPL-tokens, the `FreeSellerTradeState` is generated. The Auction House can then change the sale price to match a matching bid that is greater than 0. This allows the Auction House to do complicated order matching to find the best price for the seller and the marketplaces can write their own custom logic to do this order matching.

## Auctioneer

> In this page, we have discussed a lot about auctions and even went through a number of examples showcasing these auctions. But there was something common in all of these auctions: these auctions are all [**Double Auctions**](https://blogs.cornell.edu/info2040/2021/11/29/four-common-types-of-auctions/#:~:text=sealed%2Dbid%20auction.-,DOUBLE%20AUCTION,-Both%20buyers%20and): un-timed auctions where the buyers and sellers, bid and list until they find a common ground. There are several other forms of auctions such as English auctions and Dutch auctions which are Timed auctions that have become popular in the Solana ecosystem.

Reviewer: ```suggestion
All of the auctions we've seen so far have one thing in common. They are, what we call, [**Double Auctions**](https://blogs.cornell.edu/info2040/2021/11/29/four-common-types-of-auctions/#:~:text=sealed%2Dbid%20auction.-,DOUBLE%20AUCTION,-Both%20buyers%20and). That is, they are un-timed auctions where buyers and sellers, bid and list until they find a common ground.

However, there are several other forms of auctions such as English auctions and Dutch auctions which have become popular in the Solana ecosystem.
```

> The current Auction House implementation is designed with instant sales in mind and currently has no features that enable other auctions types.

Reviewer: ```suggestion
The Auction House implementation is purposefully designed with instant sales in mind and does not offer other auction types out-of-the-box.
```

Note that we're making AH simple by design, so Auctioneer programs can have as much power as possible.

> **Auctioneer** is a customized contract type, written by the user, that uses the composability pattern of Auction House to control an individual Auction House account. 

Reviewer: Could you mention in there that Metaplex also provides some Auctioneer implementations such as Timed Auctions?

> To fully enable Auctioneer to use an Auction House account's instructions, it must be explicitly delegated. 

Reviewer: ```suggestion
To enable an Auctioneer instance on an Auction House, it must first be explicitly delegated. This Auctioneer instance will then be able to intercept most of the Auction House instructions in order to inject its own custom logic. We will talk about Auctioneer in greater detail in later pages of this documentation.
```

![](https://i.imgur.com/RyZUfR9.png)

## Next steps

> In this page we have gone through the very basics of the Auction House protocol and the power it possesses. There is whole lot more Auction House is capable of and a lot more that can be configured to your liking.

Reviewer: ```suggestion
On this page, we have gone through the very basics of the Auction House protocol and the power it possesses. There is a lot more Auction House is capable of.
```

> To get our hands dirty with Auction House, we first need to dive a little deep into Accounts and the Instructions that are are a part of the Auction House program. Kindly go through these pages to get a good understanding of the above:

Reviewer: ```suggestion
We'll start by listing various libraries that can be used to get started with this program. Then we'll dive a little deeper into its Accounts and Instructions. This will tell us what data structures this program handles and how they are transformed.
```

* [Getting Started](/TODO)
* [Accounts](/TODO)
* [Instructions](/TODO)

> If you're building your own marketplace from scratch, its important to understand one can execute the sales, cancel bids and listings, and how one can keep track of all the listings, bids and sales. This is discussed in detail in these sections:

Reviewer: ```suggestion
If you're building your own marketplace from scratch, it's important to understand how one can execute sales, cancel bids and listings, as well as keep track of all the listings, bids and sales that occurred. This is discussed in detail in these sections:
```

* [Executing Sales](/TODO)
* [Cancelling Bids and Listings](/TODO)
* [Printing Receipts](/TODO)

> We briefly touched upon Auctioneer and Timed Auctions. These sections elaborate on how these can be setup and how Auctioneer adds utility and power to the already power-packed Auction House program:

Reviewer: ```suggestion
We briefly touched upon Auctioneer and Timed Auctions. These sections elaborate on how to get started with Auctioneer and explain how it adds more power to the Auction House program:
```

* `sellerFeeBasisPoints`: this represents the share the marketplace takes on all trades. For instance, if this is set to `200`, i.e. 2%, then the marketplace takes 2% of every single trade that happens on their platform.
    
## Listing and Bidding
Once we have an active Auction House, users can start listing assets and bidding on assets on the marketplace.

### Listing
When a user lists an asset, the Auction House does two things:
1. Auction House creates a **Sell Order**: in other words, creates a PDA known as the `SellerTradeState` which represents the listing of the asset. Trade States are special PDAs that are very cheap in comparison to other PDAs / Accounts. This is because these PDAs only store 1 byte of data, which is the [bump](https://solanacookbook.com/core-concepts/pdas.html#generating-pdas) of the PDA. All other information related to listings such as list price, amount of tokens, mint account etc, are hashed into the seeds of the PDA, instead of storing them inside the PDA itself, and therefore the PDA acts as a "proof of existence" for that listing, while being extremely cost efficient.

![](https://i.imgur.com/ki27Ds8.png)

> 2. Auction House also assigns the another PDA: `programAsSigner` PDA as the **Delegate**. Delegates are a feature of the Solana SPL-token program and are discussed in detail [here](https://spl.solana.com/token#authority-delegation). Delegation allows the Auction House to pull assets out of a token account when a sale goes through at a later point. This way, the asset need not be escrowed and can stay in the user's wallet up until the sale goes through.

Reviewer: ```suggestion
2. Auction House also assigns another PDA: `programAsSigner` PDA as the **Delegate**. Delegates are a feature of the Solana SPL-token program and are discussed in detail [here](https://spl.solana.com/token#authority-delegation). Delegation allows the Auction House to pull assets out of a token account when a sale goes through at a later point. This way, the asset need not be escrowed and can stay in the user's wallet up until the sale goes through.
```

![](https://i.imgur.com/aIRl7Hb.png)

### Bidding

> Similar to the case of listing, when a user places a bid, the Auction House creates a **Buy Order**, in other words, creates the `BuyerTradeState` PDA representing the bid. The bid amount (native or SPL tokens) needs to be transferred manually by the marketplace to the `BuyerEscrowAccount` PDA, which holds this amount till the sale goes through.

Reviewer: ```suggestion
Similar to the case of listing, when a user places a bid, the Auction House creates a **Buy Order**. In other words, it creates the `BuyerTradeState` PDA representing the bid. The bid amount (native or SPL tokens) needs to be transferred manually by the marketplace to the `BuyerEscrowAccount` PDA, which holds this amount till the sale goes through.
```

> * Bob places a bid of 5 SOL on asset A. In doing so, the marketplace pulls 5 SOL from Bob's wallet to the `BuyerEscrowAccount` PDA. This amount will stay here up until the sale goes through.
    
## Executing a Sale
Once we have a listing and at least one bid placed for a given asset, a trade can be completed by calling the `executeSale` instruction. 

The `executeSale` instruction is a permission-less crank: in other words, can be executed by anyone without any fee or reward. On the execution of the `executeSale` instruction, two things happen:
* The Auction House pulls the bid amount stored in the `BuyerEscrowAccount` and transfers this amount to the lister (minus Auction House fees). 
* The `programAsSigner` PDA, which the Auction House assigned as the **Delegate**, pulls the asset from the lister's wallet (more specifically, from the Token Account in the lister's wallet), and transfers the asset into the bidder's wallet, thus completing the trade.
![](https://i.imgur.com/gpAX63m.png)

> Now that we know how `executeSale` instruction works, lets discuss the two trade scenarios in which the `executeSale` instruction is executed in different ways:

Reviewer: ```suggestion
Now that we know how the `executeSale` instruction works, let's discuss the two trade scenarios in which the `executeSale` instruction is executed in different ways:
```

> 1. *"Buying" at list price*: This is the case when a user places a bid for a listed asset, at the listed amount itself. In such cases, the `executeSale` instruction and the bid (**Buy Order**) is done in the same instruction, and thus the bidder "buys" the asset.

Reviewer: ```suggestion
1. *"Buying" at list price*: This is the case when a user places a bid for a listed asset, at the listed amount itself. In such cases, the `bid` and the `executeSale` instructions are executed in the same transaction, and thus the bidder effectively "buys" the asset.
```

> Example:
> * Alice lists an asset A for 5 SOL. This creates a **Sell Order** for the asset A.
> * Bob notices the listing and places a bid of 5 SOL for the asset A. This creates a **Buy Order** for the asset A.
> * This enables the marketplace to place a bid on the asset and execute the sale in the same transaction, in practice allowing Bob to "buy" asset A.

> 2. *"Selling" at bid price*: This is the case when a user interested in an unlisted asset, places a bid on it. If the asset owner now lists the asset for the bid amount, the `executeSale` instruction and the listing (**Sell Order**) is done in the same instruction, and thus the lister "sells" the asset.

Reviewer: ```suggestion
2. *"Selling" at bid price*: This is the case when a user, interested in an unlisted asset, places a bid on it. If the asset owner now lists the asset for the bid amount, the `list` and the `executeSale` instructions are executed in the same instruction, and thus the lister effectively "sells" the asset at the requested price.
```

> Example:
> * Bob places a bid of 5 SOL for an unlisted asset A. This creates a **Buy Order** for the asset A.
> * Alice notices the bid and lists the asset A for 5 SOL. This creates a **Sell Order** for the asset A. 
> * This enables the marketplace to list the asset and execute the sale in the same transaction, in practice allowing Alice to "sell" asset A.

3. *Lister agreeing to a bid*: This is the case when the `executeSale` instruction is executed independently, after a **Buy Order** and a **Sell Order** exist for a given asset.

> Example:
> * Alice lists an asset A for 5 SOL. This creates a **Sell Order** for the asset A.

> > * Bob places a bid of 5 SOL for the asset A, unaware of Alice's listing. . This creates a **Buy Order** for the asset A.

Reviewer: ```suggestion
> * Bob places a bid of 5 SOL for asset A, unaware of Alice's listing. This creates a **Buy Order** for asset A.
```
